DeduplicateRecords based on recordKey if global index is used

kaushikd49 opened the pull request:
Fixes issue #344

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -1051,10 +1052,16 @@ public static SparkConf registerClasses(SparkConf conf) {
+        .mapToPair(record -> {
+          HoodieKey hoodieKey = record.getKey();
+          // If index used is global, then records are expected to differ in their partitionPath
+          boolean isGlobal = getIndex() != null && getIndex().isGlobal();
```
why would the index be null?

kaushikd49 replied:
Corrected this. The check had to be outside of this code block.

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -151,7 +152,7 @@ public HoodieWriteClient(JavaSparkContext jsc, HoodieWriteConfig clientConfig,
 
       // perform index loop up to get existing location of records
-      JavaRDD<HoodieRecord<T>> taggedRecords = index.tagLocation(dedupedRecords, table);
+      JavaRDD<HoodieRecord<T>> taggedRecords = getIndex().tagLocation(dedupedRecords, table);
```
lets remove `index` => `getIndex()`  refactors. its still a private variable that can be accessed within the same class.

kaushikd49 replied:
Sure, made a package private constructor for testing purpose to inject a mock index.

kaushikd49 commented:
@vinothchandar 
- Fixed the null check on index variable. Had to be moved outside of the code block.
- Removed getIndex(), instead added a package private constructor for testing, to inject a mocked index.

jianxu commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -122,6 +123,20 @@ public HoodieWriteClient(JavaSparkContext jsc, HoodieWriteConfig clientConfig,
     }
   }
 
+  @VisibleForTesting
```
The existing HoodieWriteClient constructor may use the new constructor to save duplicate code.
```
public HoodieWriteClient(JavaSparkContext jsc, HoodieWriteConfig clientConfig,
       boolean rollbackInFlight) {	       boolean rollbackInFlight) {
```

vinothchandar replied:
Good point.. @kaushikd49 can we fix that as well.. 

Defining a new package private constructor is much cleaner.. Nicely done!

kaushikd49 replied:
@vinothchandar @jianxu Existing public constructor does not take index as arg (which is a good thing since production code should not pass index), so cannot re-use the existing constructor. Also cannot refactor common code to a init() method because the instance variables are all finals. What do you guys think?

kaushikd49 replied:
Nvm got it :)

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -1051,10 +1058,16 @@ public static SparkConf registerClasses(SparkConf conf) {
+        .mapToPair(record -> {
+          HoodieKey hoodieKey = record.getKey();
+          // If index used is global, then records are expected to differ in their partitionPath
+          Object key = isIndexingGlobal ? hoodieKey.getRecordKey() : hoodieKey;
```
nit: replace `isIndexingGlobal` with `index.isGlobal()`

kaushikd49 commented:
Thanks, Squashed commits @vinothchandar
